src/main/java/org/roaringbitmap/longlong/RoaringTreeMap.java

```diff
@@ -0,0 +1,284 @@
+package org.roaringbitmap.longlong;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import org.roaringbitmap.IntIterator;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+//  this class is not thread-safe
+//@Beta
+public class RoaringTreeMap {
+	protected final NavigableMap<Integer, MutableRoaringBitmap> hiToBitmap = new TreeMap<>();
+
+	// Prevent recomputing all cardinalities when requesting consecutive ranks
+	protected boolean allValid = false;
+	protected int firstHiNotValid = Integer.MIN_VALUE;
+
+	// TODO: I would prefer not managing arrays myself
+	protected long[] sortedCumulatedCardinality = new long[0];
+	protected int[] sortedHighs = new int[0];
+
+	protected final List<MutableRoaringBitmap> linkedBitmaps = new ArrayList<>();
+
+	// Prevent indirection when writing consecutive Integers
+	private transient Map.Entry<Integer, MutableRoaringBitmap> latest = null;
+
+	// https://stackoverflow.com/questions/12772939/java-storing-two-ints-in-a-long
+	public void addLong(long id) {
+		int x = (int) (id >> 32);
+		int y = (int) id;
+
+		Map.Entry<Integer, MutableRoaringBitmap> local = latest;
+		if (local != null && local.getKey().intValue() == x) {
+			local.getValue().add(y);
+		} else {
+			MutableRoaringBitmap bitmap = hiToBitmap.get(x);
+			if (bitmap == null) {
+				bitmap = new MutableRoaringBitmap();
+				hiToBitmap.put(x, bitmap);
+			}
+			bitmap.add(y);
+			latest = new AbstractMap.SimpleImmutableEntry<>(x, bitmap);
+		}
+
+		// The cardinalities after this bucket may not be valid anymore
+		firstHiNotValid = Math.min(firstHiNotValid, x);
+	}
+
+	private long pack(int x, int y) {
+		return (((long) x) << 32) | (y & 0xffffffffL);
+	}
+
+	public long getCardinality() {
+		if (hiToBitmap.isEmpty()) {
+			return 0L;
+		}
+
+		ensureCumulatives(Integer.MAX_VALUE);
+
+		return sortedCumulatedCardinality[sortedCumulatedCardinality.length - 1];
+	}
+
+	public long select(final long j) {
```

Comment:
Again, the performance hit comes from the NavigableMap iteration. I would need to provide more figures, my original implementation relies on FastUtil, I quickly ported here a version not dependning on FastUtil. TreeMap iteration may be faster but I doubt about this.